### DISCUSSION TITLE
Breaking changes to make VeriFast sound with respect to C semantics as it deviates from assembly language

### DISCUSSION THREAD

**Discussion by btj** [verifast/verifast#315] [General]
I just pushed the [first](https://github.com/verifast/verifast/commit/7fa81165fa94fbd06e2873e05464300f5b6a85ab) of a series of breaking changes to VeriFast with the goal of making VeriFast sound with respect to the fact that C is not just syntactic sugar for assembly language: C compilers perform optimizations justified by a view of C as a high-level (i.e. Java-like as opposed to assembly-like) programming language. The main aspects of this are:
- It is not sound to treat uninitialized memory as if it was regular memory with some unspecified but valid and stable value. See e.g. https://www.ralfj.de/blog/2019/07/14/uninit.html and https://godbolt.org/z/dYq1Gc9G4 .
- Pointers are not just addresses: compilers are allowed to assume that pointers derived from distinct allocations are distinct, even if they happen to point to the same address in memory. This is known as pointer provenance.
- It might not be sound to assume that struct padding bytes preserve their value or otherwise behave like regular bytes.
- In the absence of the -fno-strict-aliasing flag, compilers are allowed to assume that pointers of different types do not alias. This is known as effective types.

These changes will unfortunately make VeriFast a bit harder to use and harder to teach. If you are impacted by this, do reach out so that we can see how we can make this as painless as possible. We can for example investigate providing options (similar to -disable_overflow_check) that trade off soundness for verification convenience. At first glance, for the current change this does not seem to be necessary but let me know if you do feel a need for this.

EDIT: I list below the commits that are part of this effort:
- Uninitialized memory:
  - Locals whose address is not taken: 6a5c7b464e91578d24b6d82c6158670b90d62c8b
  - Arrays: 7fa81165fa94fbd06e2873e05464300f5b6a85ab
  - Primitive-typed objects: 5917c642bcea07e7e50bc1ca68ad688924e10608
  - Fields: c04e125e74e9db097babd7e69d4b419261c92fe7
- Pointer provenance:
  - Pointers become provenance-address pairs: bc170f8b539710386e4c855f4e5e29dc1d741388
  - Tighten limits on pointer arithmetic: 4d1efe1cf18083ab1c0ff54af0fb88a7336cef89
  - Pointer equality tests: 6960d827e506b6eb55c8dfcf05747b7fedce8e4d
  - Allow only the active union member to be accessed: c2f856e4c88af315448f7f28477b6e6465981e94
- Effective types:
  - Preparation: sizeof((pointer)typeid(T)) == sizeof(T): c0823e910a32a71b416edc0db388cb16e11854f3
  - Restrict conversions between `chars` chunks and typed chunks: aa47493326cd3c43ee5813b29a9615d129ec4734
  - Enforce effective types for (u)intN_t accesses: 5c2987a856f9128e90a73177733e60430739e40a
  - Enforce effective types in `close_struct`: c2554e602e12b8fae8fc7c0e6fdf8c4a5ca246d3
  - Enforce effective types for T* accesses: f8f060b2ac4dd2bb6556b334467f55fe8b3ce031

UPDATE (27 August 2024): We have now resolved all VeriFast unsoundnesses relating to uninitialized memory, pointer provenance, and effective types that we are aware of. If further unsoundnesses are discovered in the future, they will be recorded as issues marked with label 'unsoundness'.

**Comment by dbeyer** [comment]
Thanks!  You could post a reference to the section of the C standard as support.

**Reply by btj** [reply]
Unfortunately, the C standard is not very clear on some of these points. Regarding uninitialized memory and padding bytes, see a discussion [here](https://htmlpreview.github.io/?https://github.com/C-memory-object-model-study-group/c-mom-sg/blob/master/notes/built_doc/cmom-0006-2021-03-08-clarifying-uninitialised-reads-v5.html). For provenance, see [here](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n3005.pdf). For effective types, see [here](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1796r0.pdf).

My intent is to err on the side of soundness by default (so that if VeriFast shows a green bar, the program will definitely run correctly on the major platforms), but to provide the option to reason in stronger models (that assign a more specific meaning to more programs) when this seems useful. Input on what those stronger models should be is welcome.

**Comment by jpdoyle** [comment]
EDIT: I realized this is only a problem if I'm using the syntactic sugar -- the following works:

```c
void maybe_change_x_uninit(int* x, bool do_something)
    /*@ requires int_(x,?v); @*/
    /*@ ensures  do_something ? *x |-> 1 : int_(x,v); @*/
    /*@ terminates; @*/
{
    if(do_something)
        *x = 1;
}

void always_change_x_uninit(int* x,bool do_something)
    /*@ requires *x |-> _; @*/
    /*@ ensures  *x |-> 1; @*/
    /*@ terminates; @*/
{
    *x = 1;
    maybe_change_x_uninit(x,do_something);
}
```

-------------------

Is there a way to express "this can be uninitialized, and it won't be changed"?

I don't have a fully realistic example on hand, but the following (non-verifying) snippet seems like a tricky edge case:

```c
void maybe_change_x_uninit(int* x, bool do_something)
    /*@ requires *x |-> _; @*/
    /*@ ensures  do_something ? *x |-> 1 : *x |-> _; @*/
    /*@ terminates; @*/
{
    if(do_something) *x = 1;
}

void always_change_x_uninit(int* x,bool do_something)
    /*@ requires *x |-> _; @*/
    /*@ ensures  *x |-> 1; @*/
    /*@ terminates; @*/
{
    *x = 1;
    maybe_change_x_uninit(x,do_something);
}
```
If I change the signature of `maybe_change_x_uninit` to require `*x |-> ?v` and ensure `*x |-> v`, then I no longer can pass an uninitialized pointer.

I'd almost like a way to "lift" functions that take unitialized heap chunks to functions that just preserve the values. This definitely doesn't work generically though:
```c
void dont_change_x_uninit(int* x)
    /*@ requires *x |-> _; @*/
    /*@ ensures  *x |-> _; @*/
    /*@ terminates; @*/
{
    *x = 1;
}
```

**Reply by btj** [reply]
Yes; this requires explicitly mentioning the underscore predicate for now. A possible syntactic sugar could be `*x |-?-> ?v`.

**Reply by btj** [reply]
FYI: The `LHS |-?-> RHS` syntax has now been implemented.
